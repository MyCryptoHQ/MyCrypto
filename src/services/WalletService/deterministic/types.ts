import BigNumber from 'bignumber.js';
import { ValuesType } from 'utility-types';

import { TAddress, DPathFormat, Network, ExtendedAsset, ExtendedAddressBook } from '@types';

import DeterministicWalletReducer from './reducer';
import { Wallet } from '..';

export type TDWActionError = ValuesType<typeof DeterministicWalletReducer.errorCodes>;

export interface DWAccountDisplay {
  address: TAddress;
  pathItem: {
    baseDPath: DPath;
    path: string;
    index: number;
  };
  balance: BigNumber | undefined;
  isFreshAddress?: boolean;
}

export interface ExtendedDPath extends DPath {
  offset: number;
  numOfAddresses: number;
}

export interface DeterministicWalletState {
  isInit: boolean;
  isConnected: boolean;
  isConnecting: boolean;
  isGettingAccounts: boolean;
  detectedChainId?: number | undefined;
  asset?: ExtendedAsset;
  queuedAccounts: DWAccountDisplay[];
  finishedAccounts: DWAccountDisplay[];
  customDPaths: ExtendedDPath[];
  session: Wallet | undefined;
  promptConnectionRetry: boolean;
  completed: boolean;
  errors: TDWActionError[];
}

export interface IUseDeterministicWallet {
  state: DeterministicWalletState;
  requestConnection(
    network: Network,
    asset: ExtendedAsset,
    mnemonicPhrase?: string,
    pass?: string
  ): void;
  updateAsset(asset: ExtendedAsset): void;
  addDPaths(dpaths: ExtendedDPath[]): void;
  generateFreshAddress(defaultDPath: ExtendedDPath): boolean;
}

export interface IDeterministicWalletService {
  init(walletId: DPathFormat, asset: ExtendedAsset, phrase: string, pass: string): void;
  getAccounts(session: Wallet, dpath: ExtendedDPath[]): void;
  handleAccountsQueue(accounts: DWAccountDisplay[], network: Network, asset: ExtendedAsset): void;
  triggerComplete(): void;
}
