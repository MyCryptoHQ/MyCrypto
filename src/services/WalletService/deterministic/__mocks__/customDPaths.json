[
  {
    "label": "Test Ledger Live (ETH)",
    "value": "m/44'/60'/0'/0/0",
    "isHardened": true,
    "offset": 0,
    "numOfAddresses": 5
  }
]
