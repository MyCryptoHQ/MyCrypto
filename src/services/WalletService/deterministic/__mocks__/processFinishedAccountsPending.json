{
  "newGapItems": [
    { "label": "Default (ETH)", "value": "m/44'/60'/0'/0", "offset": 4, "numOfAddresses": 2 },
    { "label": "Ledger (ETH)", "value": "m/44'/60'/0'", "offset": 4, "numOfAddresses": 6 },
    { "label": "Ledger (ETC)", "value": "m/44'/60'/160720'/0'", "offset": 4, "numOfAddresses": 2 },
    { "label": "Testnet (ETH)", "value": "m/44'/1'/0'/0", "offset": 4, "numOfAddresses": 2 }
  ],
  "customDPathItems": []
}
