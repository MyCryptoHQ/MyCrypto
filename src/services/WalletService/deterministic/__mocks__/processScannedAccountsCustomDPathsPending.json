{
  "newGapItems": [],
  "customDPathItems": [
    {
      "name": "Test Ledger Live (ETH)",
      "path": "m/44'/60'/0'/0/0",
      "isHardened": true,
      "offset": 0,
      "numOfAddresses": 5
    }
  ]
}
