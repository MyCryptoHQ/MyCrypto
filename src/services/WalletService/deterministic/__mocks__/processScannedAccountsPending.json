{
  "newGapItems": [
    {
      "name": "Default (ETH)",
      "path": "m/44'/60'/0'/0",
      "offset": 4,
      "numOfAddresses": 2
    },
    {
      "name": "Ledger (ETH)",
      "path": "m/44'/60'/0'",
      "offset": 4,
      "numOfAddresses": 6
    },
    {
      "name": "Ledger (ETC)",
      "path": "m/44'/60'/160720'/0'",
      "offset": 4,
      "numOfAddresses": 2
    },
    {
      "name": "Testnet (ETH)",
      "path": "m/44'/1'/0'/0",
      "offset": 4,
      "numOfAddresses": 2
    }
  ],
  "customDPathItems": []
}
