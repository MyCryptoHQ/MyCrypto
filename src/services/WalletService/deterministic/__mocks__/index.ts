import { default as customDPaths } from './customDPaths.json';
import { default as customDPathsDone } from './customDPathsDone.json';
import { default as customDPathsPending } from './customDPathsPending.json';
import { default as finishedAccountsDone } from './finishedAccountsDone.json';
import { default as finishedAccountsPending } from './finishedAccountsPending.json';
import { default as processFinishedAccountsCustomDPathsDone } from './processFinishedAccountsCustomDPathsDone.json';
import { default as processFinishedAccountsCustomDPathsPending } from './processFinishedAccountsCustomDPathsPending.json';
import { default as processFinishedAccountsDone } from './processFinishedAccountsDone.json';
import { default as processFinishedAccountsPending } from './processFinishedAccountsPending.json';

export const fixtures = {
  customDPaths,
  customDPathsPending,
  customDPathsDone,
  finishedAccountsPending,
  finishedAccountsDone,
  processFinishedAccountsCustomDPathsDone,
  processFinishedAccountsCustomDPathsPending,
  processFinishedAccountsDone,
  processFinishedAccountsPending
};
