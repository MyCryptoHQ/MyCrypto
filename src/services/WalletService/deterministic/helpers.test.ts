import Bignumber from 'bignumber.js';
import { clone } from 'ramda';

import { TAddress } from '@types';

// eslint-disable-next-line jest/no-mocks-import
import { fixtures } from './__mocks__';
import {
  findFinishedZeroBalanceAccounts,
  processFinishedAccounts as process,
  sortAccountDisplayItems
} from './helpers';
import { DWAccountDisplay } from './types';

interface TInputToSerialize {
  address: string;
  pathItem: {
    path: string;
    baseDPath: {
      label: string;
      value: string;
      offset: number;
      numOfAddresses: number;
    };
    index: number;
  };
  balance: string;
}

const serialize = (item: TInputToSerialize): DWAccountDisplay => {
  return clone({
    ...item,
    address: item.address as TAddress,
    balance: new Bignumber(item.balance)
  });
};

describe('deterministic wallets helpers - processFinishedAccounts', () => {
  const GAP_TO_USE = 5;
  it('processes done finishedAccounts and empty customdpaths to done', () => {
    const output = process(fixtures.finishedAccountsDone.map(serialize), [], GAP_TO_USE);
    expect(output).toStrictEqual(fixtures.processFinishedAccountsDone);
  });
  it('processes pending finishedAccounts and empty customdpaths to pending', () => {
    const output = process(fixtures.finishedAccountsPending.map(serialize), [], GAP_TO_USE);
    expect(output).toStrictEqual(fixtures.processFinishedAccountsPending);
  });
  it('processes done finishedAccounts w/done customdpaths to done', () => {
    const output = process(
      fixtures.customDPathsDone.map(serialize),
      fixtures.customDPaths,
      GAP_TO_USE
    );
    expect(output).toStrictEqual(fixtures.processFinishedAccountsCustomDPathsDone);
  });
  it('processes pending finishedAccounts w/pending customdpaths to pending', () => {
    const output = process(
      fixtures.customDPathsPending.map(serialize),
      fixtures.customDPaths,
      GAP_TO_USE
    );
    expect(output).toStrictEqual(fixtures.processFinishedAccountsCustomDPathsPending);
  });
});

describe('deterministic wallets helpers - sortAccountDisplayItems', () => {
  it('sorts account display items correctly based on index', () => {
    const outputArr = sortAccountDisplayItems(fixtures.finishedAccountsDone);
    expect(outputArr).toMatchSnapshot();
  });
});

describe('deterministic wallets helpers - findFinishedZeroBalanceAccounts', () => {
  it('returns first account display item with 0 balance', () => {
    const outputArr = findFinishedZeroBalanceAccounts(fixtures.finishedAccountsDone);
    expect(outputArr).toMatchSnapshot();
  });
});
