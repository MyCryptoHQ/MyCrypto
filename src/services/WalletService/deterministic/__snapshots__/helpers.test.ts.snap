// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deterministic wallets helpers - findFinishedZeroBalanceAccounts returns first account display item with 0 balance 1`] = `
Array [
  Object {
    "address": "0xDDd23cdC6474751C217F625033bDa2AB72A49244",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 0,
      "path": "m/44'/60'/160720'/0'/0",
    },
  },
  Object {
    "address": "0x91DD690A5A87aAEd42B20826Cd442A0f371502d6",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/1'/0'/0",
      },
      "index": 0,
      "path": "m/44'/1'/0'/0/0",
    },
  },
  Object {
    "address": "0x73C4Ed979C3918D682E9CD7E80DFF3709E5f6d0c",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'/0",
      },
      "index": 1,
      "path": "m/44'/60'/0'/0/1",
    },
  },
  Object {
    "address": "0x09F6f9916F2677f790Dd2288a163aEe05b7bD7E9",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 1,
      "path": "m/44'/60'/160720'/0'/1",
    },
  },
  Object {
    "address": "0xBf6a683aF641AD7E939166dd2BE9A904F9721496",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/1'/0'/0",
      },
      "index": 1,
      "path": "m/44'/1'/0'/0/1",
    },
  },
  Object {
    "address": "0xf63f66c71e1Afd220Ad768c9d7Cd96d7F69B5998",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'/0",
      },
      "index": 2,
      "path": "m/44'/60'/0'/0/2",
    },
  },
  Object {
    "address": "0x04725bF6D10046C9b6C7F6f60A52c966efD06b56",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'",
      },
      "index": 2,
      "path": "m/44'/60'/0'/2",
    },
  },
  Object {
    "address": "0x1F7108949B12efD73Fd3F31EB66f319d2FF0a48D",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 2,
      "path": "m/44'/60'/160720'/0'/2",
    },
  },
  Object {
    "address": "0x2D70Fbc9e12f8c0a361AB8dAA053f27B0420AbBe",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/1'/0'/0",
      },
      "index": 2,
      "path": "m/44'/1'/0'/0/2",
    },
  },
  Object {
    "address": "0x86Efe81CE7B56b3134d35f2c93D8cb6fFb10157f",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'/0",
      },
      "index": 3,
      "path": "m/44'/60'/0'/0/3",
    },
  },
  Object {
    "address": "0xdD344f9BB3dfD7A4928361356f625be3a1Fd6876",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 3,
      "path": "m/44'/60'/160720'/0'/3",
    },
  },
  Object {
    "address": "0x20a858b9249cc929d46DCDF7eaC6fd9bff9a5825",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/1'/0'/0",
      },
      "index": 3,
      "path": "m/44'/1'/0'/0/3",
    },
  },
  Object {
    "address": "0xe333E5697aF7e2Bf6c311307cdFb309C04F6A929",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'/0",
      },
      "index": 4,
      "path": "m/44'/60'/0'/0/4",
    },
  },
  Object {
    "address": "0x7e061Ac5099de7Df87E2773c0439C92a554f843C",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 4,
      "path": "m/44'/60'/160720'/0'/4",
    },
  },
  Object {
    "address": "0x97c79DE11B0220c1d9F2Be401273AEB61F032b82",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/1'/0'/0",
      },
      "index": 4,
      "path": "m/44'/1'/0'/0/4",
    },
  },
  Object {
    "address": "0xe333E5697aF7e2Bf6c311307cdFb309C04F6A929",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/60'/0'/0",
      },
      "index": 4,
      "path": "m/44'/60'/0'/0/4",
    },
  },
  Object {
    "address": "0x7e061Ac5099de7Df87E2773c0439C92a554f843C",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 4,
      "path": "m/44'/60'/160720'/0'/4",
    },
  },
  Object {
    "address": "0x97c79DE11B0220c1d9F2Be401273AEB61F032b82",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/1'/0'/0",
      },
      "index": 4,
      "path": "m/44'/1'/0'/0/4",
    },
  },
  Object {
    "address": "0x7172D60F40E309e9027C80Ec0D6aFEC470820072",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/60'/0'/0",
      },
      "index": 5,
      "path": "m/44'/60'/0'/0/5",
    },
  },
  Object {
    "address": "0xFDE49f07d985c2EbDf77D573e5FA8A579c37aA10",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 5,
      "path": "m/44'/60'/160720'/0'/5",
    },
  },
  Object {
    "address": "0x9b19985D94e7dE8bd4cD2D2584dF8FCd284303Ca",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/1'/0'/0",
      },
      "index": 5,
      "path": "m/44'/1'/0'/0/5",
    },
  },
  Object {
    "address": "0x92060Cfedfeddb31Bf81e6078A854259cF7328eEc",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 6,
        "offset": 4,
        "value": "m/44'/60'/0'",
      },
      "index": 9,
      "path": "m/44'/60'/0'/9",
    },
  },
  Object {
    "address": "0x92060Cfedfeddb31Bf81e6078A854259cF7328eEc",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 9,
        "value": "m/44'/60'/0'",
      },
      "index": 9,
      "path": "m/44'/60'/0'/9",
    },
  },
  Object {
    "address": "0x8D73d070FF18Ea7be141449566372c2197680d54",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 9,
        "value": "m/44'/60'/0'",
      },
      "index": 10,
      "path": "m/44'/60'/0'/10",
    },
  },
  Object {
    "address": "0x60661C489061Bc191393654B214b27F3EB44f9Bc",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 9,
        "value": "m/44'/60'/0'",
      },
      "index": 11,
      "path": "m/44'/60'/0'/11",
    },
  },
  Object {
    "address": "0x6bA9E98E60f46118CEa276c02fde4713f1b90D82",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 9,
        "value": "m/44'/60'/0'",
      },
      "index": 12,
      "path": "m/44'/60'/0'/12",
    },
  },
  Object {
    "address": "0x456a8DB471b7006359FF57CF1257Fbb73A2F7760",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 9,
        "value": "m/44'/60'/0'",
      },
      "index": 13,
      "path": "m/44'/60'/0'/13",
    },
  },
]
`;

exports[`deterministic wallets helpers - sortAccountDisplayItems sorts account display items correctly based on index 1`] = `
Array [
  Object {
    "address": "0x298Fe5fDca148135442fa7fA3C24042F038D3B5a",
    "balance": "1000000",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'/0",
      },
      "index": 0,
      "path": "m/44'/60'/0'/0/0",
    },
  },
  Object {
    "address": "0xB2BB2b958aFA2e96dAb3F3Ce7162B87dAea39017",
    "balance": "1000000",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'",
      },
      "index": 0,
      "path": "m/44'/60'/0'/0",
    },
  },
  Object {
    "address": "0xDDd23cdC6474751C217F625033bDa2AB72A49244",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 0,
      "path": "m/44'/60'/160720'/0'/0",
    },
  },
  Object {
    "address": "0x91DD690A5A87aAEd42B20826Cd442A0f371502d6",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/1'/0'/0",
      },
      "index": 0,
      "path": "m/44'/1'/0'/0/0",
    },
  },
  Object {
    "address": "0x73C4Ed979C3918D682E9CD7E80DFF3709E5f6d0c",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'/0",
      },
      "index": 1,
      "path": "m/44'/60'/0'/0/1",
    },
  },
  Object {
    "address": "0x8d84C8E14Ee8B755419F6764985437905eB6176E",
    "balance": "1000000",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'",
      },
      "index": 1,
      "path": "m/44'/60'/0'/1",
    },
  },
  Object {
    "address": "0x09F6f9916F2677f790Dd2288a163aEe05b7bD7E9",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 1,
      "path": "m/44'/60'/160720'/0'/1",
    },
  },
  Object {
    "address": "0xBf6a683aF641AD7E939166dd2BE9A904F9721496",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/1'/0'/0",
      },
      "index": 1,
      "path": "m/44'/1'/0'/0/1",
    },
  },
  Object {
    "address": "0xf63f66c71e1Afd220Ad768c9d7Cd96d7F69B5998",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'/0",
      },
      "index": 2,
      "path": "m/44'/60'/0'/0/2",
    },
  },
  Object {
    "address": "0x04725bF6D10046C9b6C7F6f60A52c966efD06b56",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'",
      },
      "index": 2,
      "path": "m/44'/60'/0'/2",
    },
  },
  Object {
    "address": "0x1F7108949B12efD73Fd3F31EB66f319d2FF0a48D",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 2,
      "path": "m/44'/60'/160720'/0'/2",
    },
  },
  Object {
    "address": "0x2D70Fbc9e12f8c0a361AB8dAA053f27B0420AbBe",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/1'/0'/0",
      },
      "index": 2,
      "path": "m/44'/1'/0'/0/2",
    },
  },
  Object {
    "address": "0x86Efe81CE7B56b3134d35f2c93D8cb6fFb10157f",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'/0",
      },
      "index": 3,
      "path": "m/44'/60'/0'/0/3",
    },
  },
  Object {
    "address": "0xB7CBC4e1979E3F42cBb603E837b0af5547AF5EC4",
    "balance": "1000000",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'",
      },
      "index": 3,
      "path": "m/44'/60'/0'/3",
    },
  },
  Object {
    "address": "0xdD344f9BB3dfD7A4928361356f625be3a1Fd6876",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 3,
      "path": "m/44'/60'/160720'/0'/3",
    },
  },
  Object {
    "address": "0x20a858b9249cc929d46DCDF7eaC6fd9bff9a5825",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/1'/0'/0",
      },
      "index": 3,
      "path": "m/44'/1'/0'/0/3",
    },
  },
  Object {
    "address": "0xe333E5697aF7e2Bf6c311307cdFb309C04F6A929",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'/0",
      },
      "index": 4,
      "path": "m/44'/60'/0'/0/4",
    },
  },
  Object {
    "address": "0x255b2b3A1234561D4f32D0ED4a012b73fc32C08a",
    "balance": "1000000",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/0'",
      },
      "index": 4,
      "path": "m/44'/60'/0'/4",
    },
  },
  Object {
    "address": "0x7e061Ac5099de7Df87E2773c0439C92a554f843C",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 4,
      "path": "m/44'/60'/160720'/0'/4",
    },
  },
  Object {
    "address": "0x97c79DE11B0220c1d9F2Be401273AEB61F032b82",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 5,
        "offset": 0,
        "value": "m/44'/1'/0'/0",
      },
      "index": 4,
      "path": "m/44'/1'/0'/0/4",
    },
  },
  Object {
    "address": "0xe333E5697aF7e2Bf6c311307cdFb309C04F6A929",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/60'/0'/0",
      },
      "index": 4,
      "path": "m/44'/60'/0'/0/4",
    },
  },
  Object {
    "address": "0x255b2b3A1234561D4f32D0ED4a012b73fc32C08a",
    "balance": "1000000",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 6,
        "offset": 4,
        "value": "m/44'/60'/0'",
      },
      "index": 4,
      "path": "m/44'/60'/0'/4",
    },
  },
  Object {
    "address": "0x7e061Ac5099de7Df87E2773c0439C92a554f843C",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 4,
      "path": "m/44'/60'/160720'/0'/4",
    },
  },
  Object {
    "address": "0x97c79DE11B0220c1d9F2Be401273AEB61F032b82",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/1'/0'/0",
      },
      "index": 4,
      "path": "m/44'/1'/0'/0/4",
    },
  },
  Object {
    "address": "0x7172D60F40E309e9027C80Ec0D6aFEC470820072",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Default (ETH)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/60'/0'/0",
      },
      "index": 5,
      "path": "m/44'/60'/0'/0/5",
    },
  },
  Object {
    "address": "0xd60Fb617f1A7d0Ae7F7a1231d5350B9029471aE2",
    "balance": "15557790000000000",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 6,
        "offset": 4,
        "value": "m/44'/60'/0'",
      },
      "index": 5,
      "path": "m/44'/60'/0'/5",
    },
  },
  Object {
    "address": "0xFDE49f07d985c2EbDf77D573e5FA8A579c37aA10",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETC)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/60'/160720'/0'",
      },
      "index": 5,
      "path": "m/44'/60'/160720'/0'/5",
    },
  },
  Object {
    "address": "0x9b19985D94e7dE8bd4cD2D2584dF8FCd284303Ca",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Testnet (ETH)",
        "numOfAddresses": 2,
        "offset": 4,
        "value": "m/44'/1'/0'/0",
      },
      "index": 5,
      "path": "m/44'/1'/0'/0/5",
    },
  },
  Object {
    "address": "0x28be4A0Adddd4cf084Fa7132FaE061e1CC225436",
    "balance": "1000000",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 6,
        "offset": 4,
        "value": "m/44'/60'/0'",
      },
      "index": 6,
      "path": "m/44'/60'/0'/6",
    },
  },
  Object {
    "address": "0xba34a49E3B03a58832bc4d2ba1DF8000b1301e93",
    "balance": "1000000",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 6,
        "offset": 4,
        "value": "m/44'/60'/0'",
      },
      "index": 7,
      "path": "m/44'/60'/0'/7",
    },
  },
  Object {
    "address": "0x87DC0cddd4495Aa854BcA2521691c3DFb0f1396c",
    "balance": "1000000",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 6,
        "offset": 4,
        "value": "m/44'/60'/0'",
      },
      "index": 8,
      "path": "m/44'/60'/0'/8",
    },
  },
  Object {
    "address": "0x92060Cfedfeddb31Bf81e6078A854259cF7328eEc",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 6,
        "offset": 4,
        "value": "m/44'/60'/0'",
      },
      "index": 9,
      "path": "m/44'/60'/0'/9",
    },
  },
  Object {
    "address": "0x92060Cfedfeddb31Bf81e6078A854259cF7328eEc",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 9,
        "value": "m/44'/60'/0'",
      },
      "index": 9,
      "path": "m/44'/60'/0'/9",
    },
  },
  Object {
    "address": "0x8D73d070FF18Ea7be141449566372c2197680d54",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 9,
        "value": "m/44'/60'/0'",
      },
      "index": 10,
      "path": "m/44'/60'/0'/10",
    },
  },
  Object {
    "address": "0x60661C489061Bc191393654B214b27F3EB44f9Bc",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 9,
        "value": "m/44'/60'/0'",
      },
      "index": 11,
      "path": "m/44'/60'/0'/11",
    },
  },
  Object {
    "address": "0x6bA9E98E60f46118CEa276c02fde4713f1b90D82",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 9,
        "value": "m/44'/60'/0'",
      },
      "index": 12,
      "path": "m/44'/60'/0'/12",
    },
  },
  Object {
    "address": "0x456a8DB471b7006359FF57CF1257Fbb73A2F7760",
    "balance": "0",
    "pathItem": Object {
      "baseDPath": Object {
        "label": "Ledger (ETH)",
        "numOfAddresses": 5,
        "offset": 9,
        "value": "m/44'/60'/0'",
      },
      "index": 13,
      "path": "m/44'/60'/0'/13",
    },
  },
]
`;
